src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTableEnumerator.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// Enumerate through a chain of symbol tables in depth first order. 
+    /// Ignores nulls, walks <see cref="ReadOnlySymbolTable.Parent"/>.
+    /// Throws if there are any loops. 
+    /// </summary>
+    internal class SymbolTableEnumerator : IEnumerable<ReadOnlySymbolTable>
+    {
+        private readonly ReadOnlySymbolTable[] _tables;
+
+        public SymbolTableEnumerator(params ReadOnlySymbolTable[] tables)
+        {
+            _tables = tables;
+        }
+
+        public IEnumerator<ReadOnlySymbolTable> GetEnumerator()
+        {
+            return GetEnumeratorCore();
+        }
+
+        private IEnumerator<ReadOnlySymbolTable> GetEnumeratorCore()
+        {
+            var seen = new HashSet<ReadOnlySymbolTable>();
+
+            foreach (var table in _tables)
+            {
+                var t = table;
+                while (t != null)
+                {
+                    if (!seen.Add(t))
+                    {
+                        throw new InvalidOperationException($"Loop in symbol tables");
```

Comment:
This probably isn't an issue anymore, given that it's now `set; init;` on specifying parents. Still doesn't hurt to be defensive.

Reply:
right - definitely want to be defensive, especially around loops. 
 Still including the test to catch it.